RecoVertex/BeamSpotProducer/plugins/OnlineBeamSpotESProducer.cc

```diff
@@ -0,0 +1,102 @@
+#include "CondFormats/DataRecord/interface/BeamSpotOnlineHLTObjectsRcd.h"
+#include "CondFormats/DataRecord/interface/BeamSpotTransientObjectsRcd.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+
+#include "FWCore/Framework/interface/ESProductHost.h"
+#include "FWCore/Utilities/interface/ReusableObjectHolder.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include <memory>
+#include <iostream>
+#include <string>
+
+using namespace edm;
+
+class OnlineBeamSpotESProducer : public edm::ESProducer {
+public:
+  OnlineBeamSpotESProducer(const edm::ParameterSet& p);
+  ~OnlineBeamSpotESProducer() override;
+  std::shared_ptr<const BeamSpotObjects> produce(const BeamSpotTransientObjectsRcd&);
+  static void fillDescriptions(edm::ConfigurationDescriptions& desc);
+
+private:
+  const BeamSpotOnlineObjects* compareBS(const BeamSpotOnlineObjects* bs1, const BeamSpotOnlineObjects* bs2);
+  BeamSpotObjects* fakeBS_;
+
+  edm::ESGetToken<BeamSpotObjects, BeamSpotTransientObjectsRcd> const bsToken_;
+  edm::ESGetToken<BeamSpotOnlineObjects, BeamSpotOnlineHLTObjectsRcd> bsHLTToken_;
+  edm::ESGetToken<BeamSpotOnlineObjects, BeamSpotOnlineLegacyObjectsRcd> bsLegacyToken_;
+};
+OnlineBeamSpotESProducer::OnlineBeamSpotESProducer(const edm::ParameterSet& p) {
+  auto cc = setWhatProduced(this);
+
+  fakeBS_ = new BeamSpotOnlineObjects;
+  fakeBS_->SetBeamWidthX(0.1);
+  fakeBS_->SetBeamWidthY(0.1);
+  fakeBS_->SetSigmaZ(15.);
+  fakeBS_->SetPosition(0., 0., 0.);
+  fakeBS_->SetType(-1);
+
+  bsHLTToken_ = cc.consumesFrom<BeamSpotOnlineObjects, BeamSpotOnlineHLTObjectsRcd>();
+  bsLegacyToken_ = cc.consumesFrom<BeamSpotOnlineObjects, BeamSpotOnlineLegacyObjectsRcd>();
+}
+
+void OnlineBeamSpotESProducer::fillDescriptions(edm::ConfigurationDescriptions& desc) {
+  edm::ParameterSetDescription dsc;
+  desc.addWithDefaultLabel(dsc);
+}
+
+const BeamSpotOnlineObjects* OnlineBeamSpotESProducer::compareBS(const BeamSpotOnlineObjects* bs1,
+                                                                 const BeamSpotOnlineObjects* bs2) {
+  //Random logic so far ...
+  if (bs1->GetSigmaZ() - 0.0001 < bs2->GetSigmaZ()) {  //just temporary for debugging
+    if (bs1->GetSigmaZ() > 5.) {
+      return bs1;
+    } else {
+      return bs2;
+    }
+
+  } else {
+    if (bs2->GetSigmaZ() > 5.) {
+      return bs2;
+    } else {
+      return bs1;
+    }
+  }
+}
+OnlineBeamSpotESProducer::~OnlineBeamSpotESProducer() {
+  //delete fakeBS_;
+}
+
+std::shared_ptr<const BeamSpotObjects> OnlineBeamSpotESProducer::produce(const BeamSpotTransientObjectsRcd& iRecord) {
+  auto legacyRec = iRecord.tryToGetRecord<BeamSpotOnlineLegacyObjectsRcd>();
+  auto hltRec = iRecord.tryToGetRecord<BeamSpotOnlineHLTObjectsRcd>();
+  if (not legacyRec and not hltRec) {
+    return std::shared_ptr<const BeamSpotObjects>(&(*fakeBS_), edm::do_nothing_deleter());
```

Comment:
Trying to understand the logic here, apologies if I miss something.
The `&*` operators here seems a bit odd, why not simply cast the pointer as you do a few lines below: https://github.com/cms-sw/cmssw/pull/30965/files#diff-7a6e1eb8a02607ab4693ae679bc990f4R99 ?

Reply:
Following your suggestion gives this error:
/afs/cern.ch/work/g/gennai/Run3/BeamSpot/test/CMSSW_11_2_0_pre3/src/RecoVertex/BeamSpotProducer/plugins/OnlineBeamSpotESProducer.cc: In member function 'std::shared_ptr<const BeamSpotObjects> OnlineBeamSpotESProducer::produce(const BeamSpotTransientObjectsRcd&)':
/afs/cern.ch/work/g/gennai/Run3/BeamSpot/test/CMSSW_11_2_0_pre3/src/RecoVertex/BeamSpotProducer/plugins/OnlineBeamSpotESProducer.cc:83:59: error: could not convert '(((void)std::shared_ptr<const BeamSpotObjects>(((OnlineBeamSpotESProducer*)this)->OnlineBeamSpotESProducer::fakeBS_)), edm::do_nothing_deleter())' from 'edm::do_nothing_deleter' to 'std::shared_ptr<const BeamSpotObjects>'
     return std::shared_ptr<const BeamSpotObjects>(fakeBS_), edm::do_nothing_deleter());
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
So apparently it does not work for shared_ptr without the &*:
/afs/cern.ch/work/g/gennai/Run3/BeamSpot/test/CMSSW_11_2_0_pre3/src/RecoVertex/BeamSpotProducer/plugins/OnlineBeamSpotESProducer.cc:83:85: error: no matching function for call to 'std::shared_ptr<const BeamSpotObjects>::shared_ptr(std::shared_ptr<BeamSpotObjects>&, edm::do_nothing_deleter)'
     return std::shared_ptr<const BeamSpotObjects>(fakeBS_, edm::do_nothing_deleter());

Am I doing something wrong?

Reply:
If you use `std::shared_ptr`, this line could be just (I didn't test though)
```cpp
return fakeBS_;
```
This approach does have a minuscule cost because each copy and destruction of the `shared_ptr` increases and decreases the reference count, while it is known that the member `shared_ptr` outlives all the returned copies.

Reply:
I'm trying to understand the intent of the code. If you define `std::shared_ptr<fakeBS_ BeamSpotObjects> fakeBS_` above in as a class member, you want to return a new `shared_ptr` that points to the same data, but has a different deleter? You could also write `std::shared_ptr<const BeamSpotObjects>(fakeBS_.get(), edm::do_nothing_deleter())` in this case. I have to look into the do_nothing_deleter to understand what it changes.

Reply:
it seems to work (it compiles and the test runs w/o problems) let me know which one you guys prefer.

Reply:
> You could also write `std::shared_ptr<const BeamSpotObjects>(fakeBS_.get(), edm::do_nothing_deleter())` in this case.

In that case `shared_ptr` `fakeBS_` gives no benefit, and it would better be either a regular member of `unique_ptr`.

Reply:
> it seems to work (it compiles and the test runs w/o problems) let me know which one you guys prefer.
Just to clarify,
this was referred to the suggestion of Matti to use return fakeBS_

Reply:
Indeed, I find this to be somewhat redundant/confusing. Regular member seems to make more sense as is done here: https://github.com/cms-sw/cmssw/pull/30965/files#diff-f8286142c435ae59e1c772ac65825d91R32

Reply:
Done.